### PR TITLE
store, image: have 'snap download' use v2/refresh action=download

### PR DESCRIPTION
**Pull request by chipaca**
This moves 'snap download' (i.e. image/) to use the v2 store api, with
the ad-hoc 'download' action on the refresh endpoint.

### CODE REVIEW

File: image/helpers.go [hunk 1]
```diff
@@ -221,15 +221,18 @@ func (tsto *ToolingStore) DownloadSnap(name string, revision snap.Revision, opts
...
 
-	spec := store.SnapSpec{
+	actions := []*store.SnapAction{{
+		Action:   "download",
 		Name:     name,
-		Channel:  opts.Channel,
 		Revision: revision,
-	}
-	snap, err := sto.SnapInfo(spec, tsto.user)
+		Channel:  opts.Channel,
```
Comment (pedronis) [inline image/helpers.go]: we need code like:  
https://github.com/snapcore/snapd/blob/master/overlord/snapstate/storehelpers.go#L85

or do something in store/ itself as well

File: image/helpers.go [hunk 2]
```diff
@@ -221,15 +221,18 @@ func (tsto *ToolingStore) DownloadSnap(name string, revision snap.Revision, opts
...
+	}}
+
+	snaps, err := sto.SnapAction(context.TODO(), nil, actions, tsto.user, nil)
 	if err != nil {
 		return "", nil, fmt.Errorf("cannot find snap %q: %v", name, err)
```
Comment (pedronis) [inline image/helpers.go]: should we keep this prefix?  we will get something like  "cannot find snap ...: cannot download: ..."

File: store/errors.go [hunk 3]
```diff
@@ -119,38 +119,59 @@ type SnapActionError struct {
...
+	if nOther == 0 {
+		// at least one of nDownload, nInstall, or nRefresh is > 0
+		switch {
+		case nDownload == 0 && nRefresh == 0:
+			header = "cannot install:"
+		case nDownload == 0 && nInstall == 0:
+			header = "cannot refresh:"
+		case nDownload == 0:
+			header = "cannot refresh or install:"
+		case nRefresh == 0 && nInstall == 0:
```
Comment (pedronis) [inline store/errors.go]: nitpick but I would put this case before the `nDownload == 0` one

File: store/store.go [hunk 4]
```diff
@@ -2204,7 +2204,9 @@ func (s *Store) snapAction(ctx context.Context, currentSnaps []*CurrentSnap, act
 	}
 
 	installNum := 0
+	downloadNum := 0
```
Comment (pedronis) [inline store/store.go]: I see tests for the error struct, but not for the new code here? missing something?

File: store/store.go [hunk 5]
```diff
@@ -2217,10 +2219,19 @@ func (s *Store) snapAction(ctx context.Context, currentSnaps []*CurrentSnap, act
 		}
 
 		instanceKey := a.SnapID
-		if a.Action == "install" {
-			installNum++
-			instanceKey = fmt.Sprintf("install-%d", installNum)
-			installKeys[instanceKey] = true
+		if a.Action == "install" || a.Action == "download" {
```
Comment (bboozzoo) [inline store/store.go]: Maybe it's the right time to add something to validate `Action`, eg.
```
func isValidAction(action string) bool {
    switch action {
    case "download", "install", "refresh":
        return true
    default:
        return false
    }
}
...

if !isValidAction(a.Action) {
    return nil, fmt.Errorf("internal error: unsupported action %q", a.Action)
}
...
```

File: cmd/snap/cmd_download.go [hunk 6]
```diff
@@ -100,6 +100,9 @@ func (x *cmdDownload) Execute(args []string) error {
 	if x.Revision == "" {
 		revision = snap.R(0)
 	} else {
+		if x.Channel != "" {
```
Comment (mvo5) [inline cmd/snap/cmd_download.go]: This looks untested. Given how trivial it is probably not a big deal.

**Pull request merged**